perf/src/client.rs

```diff
@@ -0,0 +1,183 @@
+
+    info!("connecting to {} at {}", host_name, addr);
+
+    let endpoint = quinn::EndpointBuilder::default();
+
+    let (endpoint, _) = endpoint
+        .bind(&SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0))
+        .context("binding endpoint")?;
+
+    let mut cfg = quinn::ClientConfigBuilder::default();
+    cfg.protocols(&[b"perf"]);
+    let mut cfg = cfg.build();
+    let tls_cfg: &mut rustls::ClientConfig = Arc::get_mut(&mut cfg.crypto).unwrap();
+    tls_cfg
+        .dangerous()
+        .set_certificate_verifier(SkipServerVerification::new());
+
+    let quinn::NewConnection {
+        connection,
+        mut uni_streams,
+        ..
+    } = endpoint
+        .connect_with(cfg, &addr, &host_name)?
+        .await
+        .context("connecting")?;
+
+    info!("established");
+
+    tokio::spawn(async move {
+        while let Some(Ok(stream)) = uni_streams.next().await {
+            tokio::spawn(async move {
+                if let Err(e) = drain_stream(stream).await {
+                    error!("reading response failed: {:#}", e);
+                }
+            });
+        }
+    });
+
+    tokio::select! {
+        x = drive_uni(connection.clone(), opt.uni_requests, opt.request_size) => x?,
+        x = drive_bi(connection.clone(), opt.bi_requests, opt.request_size) => x?,
+        _ = tokio::signal::ctrl_c() => {
+            info!("shutting down");
+            connection.close(0u32.into(), b"interrupted");
+        }
+    }
+
+    endpoint.wait_idle().await;
+
+    // TODO: Print stats
+
+    Ok(())
+}
+
+async fn drain_stream(mut stream: quinn::RecvStream) -> Result<()> {
+    #[rustfmt::skip]
+    let mut bufs = [
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+        Bytes::new(), Bytes::new(), Bytes::new(), Bytes::new(),
+    ];
+    while stream.read_chunks(&mut bufs[..]).await?.is_some() {}
+    trace!("response finished on {}", stream.id());
+    Ok(())
+}
+
+async fn drive_uni(connection: quinn::Connection, concurrency: u64, size: u64) -> Result<()> {
+    let sem = Arc::new(Semaphore::new(concurrency as usize));
+
+    loop {
+        let permit = sem.clone().acquire_owned().await.unwrap();
+        let send = connection.open_uni().await?;
+        trace!("sending request on {}", send.id());
+        tokio::spawn(async move {
+            if let Err(e) = request_uni(send, size).await {
+                error!("sending request failed: {:#}", e);
+            }
+            drop(permit);
+        });
+    }
+}
+
+async fn request_uni(mut send: quinn::SendStream, size: u64) -> Result<()> {
```

Comment:
This as well as `request_bi` will require another parameter to specify how much data to upload (and an implementation of it).